docs/proposals/202303-poll-based-binding.md

```diff
@@ -0,0 +1,233 @@
+and finalizes authentication and API selection in his local browser.
+
+4. kubectl `bind` regularly polls at every `pollInterval`.
+While the user process is ongoing in the user browser session
+the kube-bind backend returns an HTTP 403 status code.
+Once the user process is finalized the kube-bind backend returns an HTTP 200 status code.
+The response body includes the `kubebindv1alpha1.BindingResponse` object.
+
+kubectl `bind` plugin blocks until authentication and API selection is finalized.
+
+The following figure illustrates the poll based approach:
+
+![](poll-based.png)
+Link: https://whimsical.com/kube-bind-auth-flow-FYCNPvj5m27pKUDVYxVRGy
+
+### Backend conventions
+
+#### Existing `oauth2CodeGrant` protocol
+
+The existing `oauth2CodeGrant` authentication method implies the following protocol conventions:
+
+- **MUST** client invokes an authentication URL given by the `authenticatedURL`
+- **MUST** The authentication URL includes mutually exclusively either 1. a localhost callback port parameter `p`
+or 2. an explicit callback URL parameter `u`.
+In the case of a given localhost port parameter former case a `http://localhost:<p>/callback` callback URL is assumed.
+- **MUST** The authentication URL includes a session ID `s` that is generated on client side.
+- **MUST** The authentication URL includes a cluster ID `c` that is generated on client side.
+
+#### Proposed `oauth2CodeGrantPoll` protocol
+
+The proposed `oauth2CodeGrantPoll` introduces the following changes:
+
+1. The callback URL is removed, it is unneeded.
+2. Both, the session ID and cluster ID generation are now in the responsibility of the backend, not the client.
+3. An additional ephemeral session secret is generated on the backend side.
+
+- **MUST** client invokes a HTTP POST request against the `sessionURL`.
+The backend returns a `kubebindv1alpha1.Oauth2CodeGrantPollSession` response with the following attributes:
+```go
+// Oauth2CodeGrantPollSession is a non-CRUD resource that is returned by the server before
+// authentication.
+// It specifies the necessary session attributes and secrets
+// for a polling based OAuth2 Code Grant backed binding process.
+type Oauth2CodeGrantPollSession struct {
+  metav1.TypeMeta `json:",inline"`
+
+  // sessionID is a unique identifier for this binding session. It is valid until authentication,
+  // resource selection, and binding is finalized.
+  //
+  // +required
+  // +kubebuilder:validation:Required
+  // +kubebuilder:validation:MinLength=1
+  SessionID string `json:"sessionID"`
+
+  // clusterID is a unique identifier for the cluster where resources are bound to.
+  // The cluster ID is valid beyond the authentication, resource selection, and binding process.
+  //
+  // +required
+  // +kubebuilder:validation:Required
+  // +kubebuilder:validation:MinLength=1
+  ClusterID string `json:"clusterID"`
+
+  // sessionSecret is an ephemeral secret that is valid for this session only.
```

Comment:
Is there an advantage to the back-end generating the session secret?

Reply:
yes, as it is the only trusted source for the secret. clients could be compromised.